web/src/message_view.js

```diff
@@ -153,7 +161,27 @@ function create_and_update_message_list(filter, id_info, opts) {
             maybe_add_local_messages({
                 id_info,
                 msg_data,
+                msg_data_is_prepopulated,
             });
+
+            if (msg_data_is_prepopulated && !id_info.local_select_id) {
+                // If we don't have a message to select from the cached
+                // messages of the narrow, try again by populating from
+                // all_message_data.
+                //
+                // TODO: Try other super sets of messages to populate from.
+                msg_data_is_prepopulated = false;
+                msg_data = new MessageListData({
+                    filter,
+                    excludes_muted_topics,
+                });
+
+                maybe_add_local_messages({
+                    id_info,
+                    msg_data,
+                    msg_data_is_prepopulated,
+                });
+            }
```

Comment:
I think rather than having extra logic here, I was thinking that perhaps in `load_local_messages` (or in some wrapping function roughly where it is in its callers) would be where we'd put a loop over all `message_list_data` objects to find the one that has:
- A filter that's a superset of ours, with preference for an exact match.
- The widest range of matching message IDs matching our filter around our expected anchor. We'd need to be careful to do this comparison carefully -- `all_messages_data` is likely to have the widest range of message IDs always, and we want for, say, a `channel` filter that has the exact same range of message history (but not the very latest message overall, because it's in a different channel)

Not sure if it'll work nicely.

Reply:
Yeah, we can do this in next iteration.

Reply:
I find the logic in this function hard to verify: probably it's OK to test-deploy to start trying to find bugs, so I'll do that, but I think we'll either want to do that refactor or my spend time figuring out how to check it before we merge.

Reply:
Yeah, we need to simplify things here. I tried to spend some time on this today but didn't come up with a very good solution. I will circle back to it later today or tomorrow.

Reply:
Updated the PR with this.